compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp

```diff
@@ -761,6 +762,20 @@ static LogicalResult setTransformDialectConfig(func::FuncOp entryPoint,
   return setTranslationInfo(entryPoint, translationInfo);
 }
 
+static bool isMatvecLike(linalg::LinalgOp linalgOp) {
+  // TODO: Allow for matvec with fused dequantization.
+  if (!linalg::isaContractionOpInterface(linalgOp))
+    return false;
+
+  if (linalgOp.getNumParallelLoops() != 2)
+    return false;
+
+  if (linalgOp.getNumReductionLoops() > 2)
```

Comment:
Can you explain this constraint a bit? It seems somewhat arbitrary to me and is trying to match the class of grouped matvecs too specifically. I would expect to see a check for
 - Reduction loops are inner most
 - Either the lhs or the rhs is a matrix while the other is a vector
 - If it's specifically matvec, then we should be reducing along rows of the matrix

Reply:
This check is certainly suboptimal. I left a few TODOs in the code and an explanation in the commit message -- I wanted to relax this to more matvec types in the next revision.

The logic you proposed makes sense, but I found it nontrivial to implement all of these at once and actually evaluate on gfx940. For example with the dequant version we would be dealing with a 1-, 2-, and 3-d op shapes at the same time and it wasn't immediately clear for me how to tell if the 2d matrix is the 'vector' here. Similarly this should probably work with dynamic parallel dimensions but I didn't have the cycles to handle that and actually test e2e.

If there's something simple we could use to make it less arbitrary, I'm definitely open to ideas, but otherwise I would like to start with something constrained but tested and then relax it in a couple of increments.

Reply:
I was actually thinking of the opposite. Right now this is checking whether is it matmul-like (parallel loops == 2, reduction loop <= 2, contraction). SPIR-V has this
https://github.com/openxla/iree/blob/a910e74f0f820a0123f6bc2fb6e96cb27fd4f9ec/compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp#L56
Which got duplicated here
https://github.com/openxla/iree/blob/a910e74f0f820a0123f6bc2fb6e96cb27fd4f9ec/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp#L287

Maybe factor it out to a common place and make this utility check `isaContractionOpInterface && !isMatmulOrBatchMatmul`?

Reply:
What do you mean by 'the opposite'? This is actually derived from `isMatmulOrBatchMatmul`, but I didn't want to call it directly because in this pipeline we shouldn't care about the exact math in the innermost loop, so eventually I want to drop the `isContraction...` check too.

Reply:
Sorry was unclear, by the opposite I meant that this utility isn't constrained enough for matvec. Decoupling from the body makes sense, so probably what you want is `mlir::linalg::inferContractionDims` which will tell you how to interpret the dimensions and there would only be one loop range between `m` and `n` that is non-unit.

If you don't mind matching matmuls also, then just changing the name of the utility is fine (sorry if being a pain).

Reply:
I updated the code to check for the unit parallel bound. PTAL.